src/ui/server/controllers/insurance/account/sign-in/enter-code/index.ts

```diff
@@ -35,20 +43,80 @@ export const PAGE_CONTENT_STRINGS = PAGES.INSURANCE.ACCOUNT.SIGN_IN.ENTER_CODE;
  * @param {Express.Response} Express response
  * @returns {Express.Response.render} Enter code page
  */
-export const get = (req: Request, res: Response) =>
-  res.render(TEMPLATE, {
+export const get = (req: Request, res: Response) => {
+  // TODO: move to user.accountId
+  if (!req.session.accountId) {
+    return res.redirect(SIGN_IN_ROOT);
+  }
+
+  return res.render(TEMPLATE, {
     ...insuranceCorePageVariables({
       PAGE_CONTENT_STRINGS,
       BACK_LINK: req.headers.referer,
     }),
     ...PAGE_VARIABLES,
   });
+};
 
 /**
  * post
- * Temporary redirect to the dashboard route.
+ * Validate sign in code
  * @param {Express.Request} Express request
  * @param {Express.Response} Express response
- * @returns {Express.Response.redirect} Next part of the flow
+ * @returns {Express.Response.redirect} Next part of the flow or validation errors
  */
-export const post = (req: Request, res: Response) => res.redirect(DASHBOARD);
+export const post = async (req: Request, res: Response) => {
+  try {
+    // TODO: move to user.accountId
+    if (!req.session.accountId) {
+      return res.redirect(SIGN_IN_ROOT);
+    }
+
+    const securityCode = req.body[FIELD_ID];
+
+    let validationErrors = generateValidationErrors(req.body);
+
+    if (validationErrors) {
+      return res.render(TEMPLATE, {
+        ...insuranceCorePageVariables({
+          PAGE_CONTENT_STRINGS,
+          BACK_LINK: req.headers.referer,
+        }),
+        ...PAGE_VARIABLES,
+        submittedValues: req.body,
+        validationErrors,
+      });
+    }
+
+    const response = await api.keystone.account.verifyAccountSignInCode(req.session.accountId, securityCode);
+
+    // valid sign in code - update the session and redirect to the dashboard
+    if (response.success) {
+      req.session.user = {
+        id: response.accountId,
+        firstName: response.firstName,
+        lastName: response.lastName,
+        token: response.token,
+        expires: response.expires,
+      };
+
+      return res.redirect(DASHBOARD);
+    }
+
+    // incorrect sign in code - force validation errors by mimicking empty form submission)
+    validationErrors = securityCodeValidationErrors({}, {});
```

Comment:
rather than passing empty objects, could we not have a check in `securityCodeValidationErrors` to see if they are passed and if not, return an error

Reply:
I can't see a better way to do this - in this scenario, the code has been passed in the `req.body`, but the API says it's invalid (no `response.success`). So we need to render the validation error - but the validation rules only have "if a code is not submitted" checks - which is all the UI should care about.

I think the only other way to do this would be to pass a new flag to the validation rule e.g `forceError=true`, but I think passing an empty object is cleaner. What do you think?